src/api/bulk2.ts

```diff
@@ -0,0 +1,1000 @@
+  async uploadData(input: string | Record[] | Readable): Promise<void> {
+    await this.jobData.execute(input);
+  }
+
+  async getAllResults(): Promise<IngestJobV2Results<S>> {
+    const [
+      successfulResults,
+      failedResults,
+      unprocessedRecords,
+    ] = await Promise.all([
+      this.getSuccessfulResults(),
+      this.getFailedResults(),
+      this.getUnprocessedRecords(),
+    ]);
+    return { successfulResults, failedResults, unprocessedRecords };
+  }
+
+  /**
+   * Close opened job
+   */
+  async close(): Promise<void> {
+    try {
+      const state: JobStateV2 = 'UploadComplete';
+      this.jobInfo = await this.createIngestRequest<JobInfoV2>({
+        method: 'PATCH',
+        path: `/${this.jobInfo.id}`,
+        body: JSON.stringify({ state }),
+        headers: { 'Content-Type': 'application/json; charset=utf-8' },
+        responseType: 'application/json',
+      });
+      this.logger.debug(this.jobInfo);
+      this.emit('uploadcomplete');
+    } catch (err) {
+      this.emit('error', err);
+      throw err;
+    }
+  }
+
+  /**
+   * Set the status to abort
+   */
+  async abort(): Promise<void> {
+    try {
+      const state: JobStateV2 = 'Aborted';
+      this.jobInfo = await this.createIngestRequest<JobInfoV2>({
+        method: 'PATCH',
+        path: `/${this.jobInfo.id}`,
+        body: JSON.stringify({ state }),
+        headers: { 'Content-Type': 'application/json; charset=utf-8' },
+        responseType: 'application/json',
+      });
+      this.logger.debug(this.jobInfo);
+      this.emit('aborted');
+    } catch (err) {
+      this.emit('error', err);
+      throw err;
+    }
+  }
+
+  /**
+   * Poll for the state of the processing for the job.
+   *
+   * This method will only throw after a timeout. To capture a
+   * job failure while polling you must set a listener for the
+   * `failed` event before calling it:
+   *
+   * job.on('failed', (err) => console.error(err))
+   * await job.poll()
+   *
+   * @param interval Polling interval in milliseconds
+   * @param timeout Polling timeout in milliseconds
+   * @returns {Promise<void>} A promise that resolves when the job finishes successfully
+   */
+  async poll(
+    interval: number = this.pollingOptions.pollInterval,
+    timeout: number = this.pollingOptions.pollTimeout,
+  ): Promise<void> {
+    const jobId = getJobId(this.jobInfo);
+    const startTime = Date.now();
+
+    this.logger.debug(`Job ID: ${jobId}`);
+    this.logger.debug(
+      `Polling options: timeout:${timeout}ms | interval: ${interval}ms.`,
+    );
+
+    while (startTime + timeout > Date.now()) {
+      try {
+        const res = await this.check();
+        switch (res.state) {
+          case 'Open':
+            throw new Error('Job has not been started');
+          case 'Aborted':
+            throw new Error('Job has been aborted');
+          case 'UploadComplete':
+          case 'InProgress':
+            await delay(interval);
+            break;
```

Comment:
this is a good place for, at a minimum, a logger statement about the progress.

I think it'd be cooler to `emit` some event, like 

```ts
numberRecordsProcessed?: number;
numberRecordsFailed?: number;
```

Imagine wanting to build a progress bar for the upload?

Reply:
added `inProgress` event that include the full response.

bulk v1 was doing the same but emitting `progress`, now both to `inProgress`.